### DISCUSSION TITLE
SWG on when Pump is not running

### DISCUSSION THREAD

**Discussion by johnny2678** [tagyoureit/nodejs-poolController#495] [Q&A]
A couple of posts on this topic recently:
https://github.com/tagyoureit/nodejs-poolController/discussions/494
https://github.com/tagyoureit/nodejs-poolController/discussions/484#discussioncomment-2959420

I believe OPs in both cases are running pure nixie builds.  I still have an OCP (Easytouch 2 4).  It's my 2nd morning live on my new pi4 build with the sequent stack and I'm 2-for-2 observing the pump powered, but not running while the SWG is on at 100% when the OCP filter circuit gets turned on in the AM on a schedule.

Fairly certain this is due to the super watchdog but I have no idea why.  The Pi stack shuts down at some point during the night, presumably by the super watchdog - I do have that configured according to the wiki.  I've removed it from my stack to see if I have the same problem tmr AM.

Now, I have noticed on this new stack that when the pi shuts down, the pump motor stops.  I could see why you might want that for a nixie build, but if njspc isn't there and an OCP is present, the OCP should just assume full control, right?   Pretty sure that's how it always was on my old rPi 3b stack. 

Not sure if that's because rs485 is running through the BAS now instead of my old jbtek usb adapter but I could always shut the pi down before on my 3b stack and the system would just continue to be controlled by the OCP - pump would stay on.

Any thoughts on why the pump would stop running when the pi shuts down?  Can't really grab a replay because it happens after the pi shuts down.

**Comment by johnny2678** [comment]
So, it's not the super watchdog

I got to the pad one minute past the scheduled start time this morning.  When I got there, the pump was not running, but the filter circuit was on - swg had power.  njspc had gone non-responsive, not sure if the pi shutdown or just the process.  I had just checked dashPanel at 7:50 and it was working.

here's the interesting thing.  The pump panel was under local control - like it didn't know it had a controller connected via rs485, either njspc or the EasyTouch.

![image](https://user-images.githubusercontent.com/8219743/175027655-da604761-389b-4614-aa8f-5934dc5298df.png)

As soon as I pulled power to the pi, the OCP took over, the pump panel reverted to what it normally says - "Display Not Active", and the pump motor started.

![image](https://user-images.githubusercontent.com/8219743/175027874-f6aa2f62-029a-483f-bb02-51e7e6c3e456.png)

Once the pump was powered on, i plugged the Pi back in and the stack powered up and I can get rs485 data and control via dashpanel.

I've been running this stack for years but this is only the 3rd day that I've been running it with rs485 running through the MEGA-BAS, and this has happened all three days.  I'm supposed to be set on RS485 Pass-Through, yes?

![image](https://user-images.githubusercontent.com/8219743/175028592-1e2b85a4-63f3-4467-8c16-17ab17562a2f.png)

Does any of this make sense?  I guess tomorrow I will try and start a replay just prior to schedule start at 8AM.

**Reply by tagyoureit** [reply]
> As soon as I pulled power to the pi, the OCP took over, the pump panel reverted to what it normally says - "Display Not Active", and the pump motor started.

Do attach a replay, but this should not be happening.  OCP should be in control of the pump 100% of the time and njsPC should just be snooping on that bus traffic.  How do you have the RS485 wired up between the OCP, Sequent/PI and the Pump?  

> I'm supposed to be set on RS485 Pass-Through, yes?

yes

> njspc had gone non-responsive, not sure if the pi shutdown or just the process.

You need to track this down.  Can you remote into the Pi at this point?  Is there anything in the logs?

**Reply by johnny2678** [reply]
> You need to track this down.

yep, planning on being on the pad tomorrow a few minutes before scheduled start tomorrow (8AM) to observe.

**Comment by johnny2678** [comment]
Stayed up late thinking about this.  Then it hit me that I have a script attached to the sequent relay hat button.  Falling edge on GPIO 26 performs a system shutdown.  Nice to have at the outdoor enclosure when I need a clean shutdown.

```
Shutting down the Raspberry Pi by turning off the power can result in SD Card failure. To prevent this, a shutdown command needs to be used before power cut-off. But this requires a monitor, keyboard and mouse connected to the Pi. 

A momentary on push button installed at the edge of the board provides a convenient way to shut down the Raspberry Pi. The button is routed to pin 37 (GPIO 26). You need to write a script which monitors this pin, and if pressed for more than a desired time, issues the shut-down command.
```

Only, for some reason it seems that I was getting random falling edges on GPIO.  Not sure why they would always coincide with the pool schedule starting at 8am.  But I disabled that script and pool started fine this AM.

Maybe something could be built into the REM device type for the sequent relay boards that have this button?  Definitely leaving the script off for now.  Not worth it.  

Couple of unanswered questions still:
1.  Why was I getting a falling edge on GPIO 26 when nothing is hooked up to it except the relay button?
2. Why would shutting down njspc cause the pump to shut down when I have an EasyTouch?  I guess I could still test this one.  Might play around with showdowns/brown-outs today and see what happens.

**Reply by tagyoureit** [reply]
> Why was I getting a falling edge on GPIO 26 when nothing is hooked up to it except the relay button?
Does REM have any assignments at all for GPIO26?    Can you upload the script you were using?

> Why would shutting down njspc cause the pump to shut down when I have an EasyTouch? I guess I could still test this one. Might play around with showdowns/brown-outs today and see what happens.

You should be able to replicate this by just shutting down the Pi, right?  Keep in mind the pump may still run for up to 10 mins before it resets itself to be a standalone pump.  But the only thing that comes to mind is wiring.  If the Pi is acting in any way as a break in the wiring when it is off then I could see this happening.  

If you really get stuck, you could use your spare Pi and load up another copy of njsPC/dp and hook it up to the rs485 and see what happens with the pump messages when the primary pi is shutdown.

**Reply by tagyoureit** [reply]
Another to try, to really ensure that it's the script, is to replace the reboot command with a timestamp written to a file.  `echo "$(date)" >> reboottime.txt`.  Maybe there is something else besides GPIO26 that's causing a reboot (power surge/issue?) and this would help understand that component as well.

**Reply by rstrouse** [reply]
I haven't messed with this much but do you need to enable the internal pull-down on the pi to ensure GPIO 26 does not float?  If that is the case and you didn't pull the GPIO up or down then this condition might occur.

**Reply by johnny2678** [reply]
I have no doubt that there is something wrong with my script logic somewhere.  It was spliced together from internet research and seemed to work in testing but not worth using now that I know it can cause problems.

Posting here in case you see something obvious

```
import time
import RPi.GPIO as GPIO #Python Package Reference: https://pypi.org/project/RPi.GPIO/
from gpiozero import Buzzer

# Pin definition
reset_shutdown_pin = 26
buzzer = Buzzer(23)

# Suppress warnings
GPIO.setwarnings(False)

# Use "GPIO" pin numbering
GPIO.setmode(GPIO.BCM)

# Use built-in internal pullup resistor so the pin is not floating
# if using a momentary push button without a resistor.
#GPIO.setup(reset_shutdown_pin, GPIO.IN, pull_up_down=GPIO.PUD_UP)

# Use Qwiic pHAT's pullup resistor so that the pin is not floating
GPIO.setup(reset_shutdown_pin, GPIO.IN)

# modular function to restart Pi
def restart():
    print("restarting Pi")
    command = "/usr/bin/sudo /sbin/shutdown -r now"
    buzzer.on()
    time.sleep(.1)
    buzzer.off()
    time.sleep(.2)
    buzzer.on()
    time.sleep(.1)
    buzzer.off()
    import subprocess
    process = subprocess.Popen(command.split(), stdout=subprocess.PIPE)
    output = process.communicate()[0]
    print(output)

# modular function to shutdown Pi
def shut_down():
    print("shutting down")
    command = "/usr/bin/sudo /sbin/shutdown -h now"
    buzzer.on()
    time.sleep(.1)
    buzzer.off()
    time.sleep(.2)
    buzzer.on()
    time.sleep(.1)
    buzzer.off()
    buzzer.on()
    time.sleep(.1)
    buzzer.off()
    time.sleep(.2)
    buzzer.on()
    time.sleep(.1)
    buzzer.off()
    import subprocess
    process = subprocess.Popen(command.split(), stdout=subprocess.PIPE)
    output = process.communicate()[0]
    print(output)




while True:
    #short delay, otherwise this code will take up a lot of the Pi's processing power
    time.sleep(0.5)

    # wait for a button press with switch debounce on the falling edge so that this script
    # is not taking up too many resources in order to shutdown/reboot the Pi safely
    channel = GPIO.wait_for_edge(reset_shutdown_pin, GPIO.FALLING, bouncetime=200)

    if channel is None:
        print('Timeout occurred')
    else:
        print('Edge detected on channel', channel)

        # For troubleshooting, uncomment this line to output button status on command line
        #print('GPIO state is = ', GPIO.input(reset_shutdown_pin))
        counter = 0

        while GPIO.input(reset_shutdown_pin) == False:
            # For troubleshooting, uncomment this line to view the counter. If it reaches a value above 4, we will restart.
            #print(counter)
            counter += 1
            time.sleep(0.5)

            # long button press
            if counter > 4:
                shut_down()

        #if short button press, restart!
        restart()


```

**Reply by orrious** [reply]
@rstrouse is correct.  Your GPIO is floating and you need to pull it UP.  `#GPIO.setup(reset_shutdown_pin, GPIO.IN, pull_up_down=GPIO.PUD_UP)` is commented out, uncomment it.

Make sure the other side of your button is connected to GND.  The pull up resistor will pull the GPIO up / high unless the button is pressed.  When you press the button, it pulls the GPIO to GND / low.

**Comment by subcritical71** [comment]
I wanted to enable a clean way of shutting down and restarting the RPi so I installed the above script (uncommenting as suggested by @rstrouse @orrious )

When I boot the pi headless it seems to hang somewhere in the boot process. I can't get a ssh session to work and the REM and dashPanel webpages will not load. 

When I boot with the pi connected to a monitor everything works as expected including the restart/shutdown scripts.

Any idea why I could be getting this behavior?

I added the py script above to my rc.local;

```
#
# This script is executed at the end of each multiuser runlevel.
# Make sure that the script will "exit 0" on success or any other
# value on error.
#
# In order to enable or disable this script just change the execution
# bits.
#
# By default this script does nothing.

# Print the IP address
_IP=$(hostname -I) || true
if [ "$_IP" ]; then
  printf "My IP address is %s\n" "$_IP"
fi

python /home/pi/mega-bas-gpio26.py &

exit 0

```

**Reply by rstrouse** [reply]
Do you mean that if you boot to the pi with something connected to the HDMI everything runs normally.  Can you then connect using SSH or RDP to the pi from somewhere else?  Just for shucks make sure you do not have remote GPIO turned on in the Raspi Config.

**Reply by subcritical71** [reply]
Yes, if I do HDMI, Keyboard (USB), Mouse (USB) everything works as expected (remote SSH, restart/shutdown script). If I go headless, nothing. I did try to install the script about 2 weeks ago before I started over with the RPi and it did the same thing but I chalked it up to me doing something wrong. I did disable remote GPIO as you suggested (not sure if it was on or off as the interface didn't tell me). I'll try the script again later today.

**Reply by tagyoureit** [reply]
If you have RasPi booting to desktop, it will not load unless you force it to override HDMI detection.  See [this](https://raspberrypi.stackexchange.com/questions/109279/pi-4b-wont-boot-without-hdmi-cable-plugged-to-monitor).

**Reply by subcritical71** [reply]
Looks like I've tracked down my issue. It seems I may have had a bad 5V power supply. It worked when hooked up to an external monitor because I had all the I/Os disconnected. I installed a new power supply and everything is working as expected. Man did that take a bit to figure out!
